File: enterprise/web-frontend/modules/baserow_enterprise/licenseTypes.js
import EnterpriseFeaturesObject from '@baserow_enterprise/features'
import PremiumFeaturesObject from '@baserow_premium/features'
import { LicenseType } from '@baserow_premium/licenseTypes'

const commonAdvancedFeatures = [
  // core
  PremiumFeaturesObject.PREMIUM,
  EnterpriseFeaturesObject.RBAC,
  EnterpriseFeaturesObject.TEAMS,
  EnterpriseFeaturesObject.AUDIT_LOG,
  // database
  EnterpriseFeaturesObject.DATA_SYNC,
  EnterpriseFeaturesObject.ADVANCED_WEBHOOKS,
  EnterpriseFeaturesObject.FIELD_LEVEL_PERMISSIONS,
  EnterpriseFeaturesObject.DATE_DEPENDENCY,
  // WAB
  EnterpriseFeaturesObject.BUILDER_SSO,
  EnterpriseFeaturesObject.BUILDER_NO_BRANDING,
  EnterpriseFeaturesObject.BUILDER_FILE_INPUT,
  EnterpriseFeaturesObject.BUILDER_CUSTOM_CODE,
  EnterpriseFeaturesObject.CODE_RUNNER,
  EnterpriseFeaturesObject.XLS_FILE_READER,
  EnterpriseFeaturesObject.BUILDER_GROUPED_AGGREGATE_ROWS,
  // Only self-hosted
  EnterpriseFeaturesObject.SSO,
]

export class AdvancedLicenseType extends LicenseType {
  static getType() {
    return 'advanced'
  }

  getName() {
    const { $i18n } = this.app
    return $i18n.t('advanced.license')
  }

  getLicenseBadgeColor() {
    return 'magenta'
  }

  getFeaturesDescription() {
    const { $i18n } = this.app
    return [
      {
        name: $i18n.t('license.premiumFeatureName'),
        enabled: true,
      },
      {
        name: $i18n.t('license.advancedFeatureName'),
        enabled: true,
      },
      {
        name: $i18n.t('license.enterpriseFeatureName'),
        enabled: false,
      },
      {
        name: $i18n.t('license.supportFeatureName'),
        enabled: false,
      },
    ]
  }

  getFeatures() {
    return [...commonAdvancedFeatures, EnterpriseFeaturesObject.SUPPORT]
  }

  getTopSidebarTooltip() {
    const { $i18n } = this.app
    return $i18n.t('enterprise.sidebarTooltip')
  }

  showInTopSidebarWhenActive() {
    return true
  }

  getOrder() {
    return 75
  }

  getSeatsManuallyAssigned() {
    return false
  }

  getLicenseDescription(license) {
    const { $i18n } = this.app
    return $i18n.t('enterprise.licenseDescription')
  }

  getLicenseSeatOverflowWarning(license) {
    const { $i18n } = this.app
    return $i18n.t('enterprise.overflowWarning')
  }

  getLicenseSeatOverflowButton(license) {
    return {
      href: 'https://baserow.io/subscriptions',
      text: this.app.$i18n.t('advanced.openSubscriptions'),
    }
  }
}

export class EnterpriseWithoutSupportLicenseType extends AdvancedLicenseType {
  static getType() {
    return 'enterprise_without_support'
  }

  getName() {
    const { $i18n } = this.app
    return $i18n.t('enterprise.license')
  }

  getFeatures() {
    return [
      ...commonAdvancedFeatures,
      EnterpriseFeaturesObject.ENTERPRISE_SETTINGS,
    ]
  }

  getFeaturesDescription() {
    const description = super.getFeaturesDescription()
    description[2].enabled = true
    return description
  }

  getOrder() {
    return 100
  }

  getLicenseBadgeColor() {
    return 'neutral'
  }

  getLicenseSeatOverflowButton(license) {
    return {
      href: 'https://baserow.io/contact-sales',
      text: this.app.$i18n.t('license.contactSalesMoreSeats'),
    }
  }
}

export class EnterpriseLicenseType extends EnterpriseWithoutSupportLicenseType {
  static getType() {
    return 'enterprise'
  }

  getFeatures() {
    return super.getFeatures().concat(EnterpriseFeaturesObject.SUPPORT)
  }

  getFeaturesDescription() {
    const description = super.getFeaturesDescription()
    description[3].enabled = true
    return description
  }

  getOrder() {
    return 101
  }
}

File: enterprise/backend/src/baserow_enterprise/license_types.py
from typing import List, Optional

from baserow.core.models import Workspace
from baserow_enterprise.features import (
    ADVANCED_WEBHOOKS,
    AUDIT_LOG,
    BUILDER_CUSTOM_CODE,
    BUILDER_FILE_INPUT,
    BUILDER_NO_BRANDING,
    BUILDER_SSO,
    CODE_RUNNER,
    DATA_SCANNER,
    DATA_SYNC,
    DATE_DEPENDENCY,
    ENTERPRISE_SETTINGS,
    FIELD_LEVEL_PERMISSIONS,
    RBAC,
    SECURE_FILE_SERVE,
    SSO,
    SUPPORT,
    TEAMS,
    XLS_FILE_READER,
)
from baserow_enterprise.role.seat_usage_calculator import (
    RoleBasedSeatUsageSummaryCalculator,
)
from baserow_premium.license.features import PREMIUM
from baserow_premium.license.models import License
from baserow_premium.license.registries import LicenseType, SeatUsageSummary

COMMON_ADVANCED_FEATURES = [
    # core
    PREMIUM,
    RBAC,
    TEAMS,
    AUDIT_LOG,
    # database
    DATA_SYNC,
    ADVANCED_WEBHOOKS,
    FIELD_LEVEL_PERMISSIONS,
    DATE_DEPENDENCY,
    # WAB
    BUILDER_SSO,
    BUILDER_NO_BRANDING,
    BUILDER_FILE_INPUT,
    BUILDER_CUSTOM_CODE,
    CODE_RUNNER,
    XLS_FILE_READER,
    # only self-hosted
    SSO,
]


class AdvancedLicenseType(LicenseType):
    """
    The advanced plan is similar to the enterprise plan. The main difference is that it
    doesn't allow branding and secure file serving. Other than that it includes all
    enterprise features. The seat limit is also enforced because it can be bought self
    served.
    """

    type = "advanced"
    order = 75
    features = [
        *COMMON_ADVANCED_FEATURES,
        SUPPORT,
    ]
    instance_wide = True
    seats_manually_assigned = False

    def get_seat_usage_summary(
        self, license_object_of_this_type: License
    ) -> SeatUsageSummary:
        return RoleBasedSeatUsageSummaryCalculator.get_seat_usage_for_entire_instance()

    def get_seat_usage_summary_for_workspace(
        self, workspace: Workspace
    ) -> Optional[SeatUsageSummary]:
        return RoleBasedSeatUsageSummaryCalculator.get_seat_usage_for_workspace(
            workspace
        )

    def get_seat_usage_summary_for_specific_users(
        self, user_ids: List[int]
    ) -> Optional[SeatUsageSummary]:
        return RoleBasedSeatUsageSummaryCalculator.get_seat_usage_for_specific_users(
            user_ids
        )

    def handle_seat_overflow(self, seats_taken: int, license_object: License):
        # We don't have to do anything because the seat limit is a soft limit. This is
        # okay for now because we'll be monitoring the usage manually.
        pass


class EnterpriseWithoutSupportLicenseType(AdvancedLicenseType):
    type = "enterprise_without_support"
    order = 100
    features = [
        *COMMON_ADVANCED_FEATURES,
        ENTERPRISE_SETTINGS,
        SECURE_FILE_SERVE,
        DATA_SCANNER,
    ]

    def handle_seat_overflow(self, seats_taken: int, license_object: License):
        # We don't have to do anything because the seat limit is a soft limit.
        pass


class EnterpriseLicenseType(EnterpriseWithoutSupportLicenseType):
    type = "enterprise"
    order = 101
    features = EnterpriseWithoutSupportLicenseType.features + [SUPPORT]
